LF-3005: Fix useUnit hook

SayakaOno opened the pull request:
**Description**

- Update `useUnit`
  - A new feature is planned to allow the user to decide whether to auto convert values ​​when units are changed. The new prop `autoConversion` is used to trigger auto conversion.
  - update `getOnChangeUnitOption` function to make it simpler
- Add prop `autoConversion` to the Unit component
- Remove `onChangeUnitOption` in `WaterUsageCalculatorModal`. (It looks like it was added to make the component work with the broken logic)

I plan to write some unit tests for the component, and there are tickets to write integration tests for components that have the unit component.

**Note:** Water Calculator for Depth crashes because I probably removed some logic that was making "% of location to be irrigated" input work. The unit component should not be used for "%", and I am going to work on replacing the input with a simple input in another ticket.

(I am working on refactoring the Unit component, and I plan to do it in 3+ PRs. The PRs I am creating are going to be merged to the "LF-2995+3002/Unit_displays" branch, and when everything is ready, the branch is going to be merged to the integration branch.)
- [x] LF-3003 Extract logic from Unit component and create a hook
- [x] LF-3004 Update Unit component to make it work as a presentation component
- [x] **LF-3005 Fix useUnit hook**

Jira link: https://lite-farm.atlassian.net/browse/LF-3005

**Type of change**

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

- [ ] Passes test case
- [ ] UI components visually reviewed on desktop view
- [ ] UI components visually reviewed on mobile view
- [x] Other
commented out line 100 (`display: none;`) in `/packages/webapp/src/components/Form/Unit/unit.module.scss` to make the hiddenInput visible and went through all instances in this [doc](https://lite-farm.atlassian.net/l/cp/gSDPpyen)
Things I tested:
      - hidden value changes correctly by updating visible value and unit
      - data retrieved from the DB is shown correctly as expected (tested after switching user's preferred farm unit system as well)

**Checklist:**

- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] The precommit and linting ran successfully
- [ ] I have added or updated language tags for text that's part of the UI
- [ ] I have added "MISSING" for all new language tags to languages I don't speak
- [ ] I have added [the GNU General Public License](https://lite-farm.atlassian.net/l/cp/BT0Dd7WW) to all new files

JimmyRowland commented on packages/webapp/src/components/Form/Unit/useUnit.jsx:
```diff
@@ -233,33 +229,34 @@ const useUnit = ({
-      // function from react-hook-form Controller's render
-      onChange(e);
+  const onChangeUnit = (e) => {
+    hookFormSetValue(displayUnitName, e);
```
https://github.com/react-hook-form/react-hook-form/blob/52925875e7a5ab41601e45c3209f6c5bda0099ff/src/logic/createFormControl.ts#L655
onChange from controller does more than just setting value. 
The safe way is to use useController hook. 
but hookFormSetValue is fine since there is no validation on unit

JimmyRowland commented on packages/webapp/src/components/Form/Unit/useUnit.jsx:
```diff
@@ -103,85 +106,84 @@ const useUnit = ({
         roundToTwoDecimal(convert(hookFormValue).from(databaseUnit).to(hookFormUnit)),
       );
       //Trigger validation
       (hookFormValue === 0 || hookFormValue > 0) && hookFormSetHiddenValue(hookFormValue);
```
why it excludes negative values?

JimmyRowland replied:
https://github.com/LiteFarmOrg/LiteFarm/pull/2079
This is one reason pull requests should be squashed per ticket.
It's very difficult to track changes in git tree

JimmyRowland commented:
![image](https://user-images.githubusercontent.com/16997894/227098765-1c083538-758f-4817-8a87-88a69a743c68.png)This 
click container and then click in ground
![image](https://user-images.githubusercontent.com/16997894/227098876-ed9f4e1b-0238-4bad-b24f-1ee0350c1083.png)

interaction is weird.
This probably doesn't matter

SayakaOno commented:
Thank you Jimmy for reviewing!
I have not done tests like that...😨 
I looked into it, and "Estimated seed required" was re-calculated every time I re-selected "In ground" option. (you might have known) I will discuss with the team how to address it. Thank you!!

Duncan-Brain commented:
@SayakaOno 

It all looks pretty darn good to me! 

My main questions are: 

1. Will there ever in the future be a need for negative values? I noticed that 0 is the minimum allowed on this component.

2. The water use calculator: since it breaks the functionality either way should we make any changes to this at all? Or save it for the other ticket you mentioned?

3. I left one small nitpicky comment which you can feel free to ignore.

Happy to approve merge onto other feature branch at any time.

SayakaOno commented on packages/webapp/src/components/Form/Unit/useUnit.jsx:
```diff
@@ -103,85 +108,84 @@ const useUnit = ({
       ? reactSelectWidth - DEFAULT_SELECT_ARROW_ICON_WIDTH
       : reactSelectWidth;
+    const displayUnit = to || hookFormDisplayUnit;
+    const values = {
```
Actually, I had a hard time coming up with this valuable name and ended up with values (useMemo's return is `cachedValue` in [the official doc](https://react.dev/reference/react/useMemo#usememo)...)
Would `initialValues` or `defaultValues` be better? (`value` is confusing...)

Duncan-Brain replied:
Haha, variable naming is so hard! I don't have any great suggestions...

In that convenience object it has units, element props like css width and visibility, event functions. Difficult to name

SayakaOno commented:
@Duncan-Brain 
Thank you for reviewing!

1. I think it's possible. Actual harvested amount was less than estimated amount, or something like that? I just fixed the logic this time, and I think some work would be needed for negative values.
2. I am not sure if I understand you correctly, but I plan to merge changes for the water calculator to the integration branch, and merge the integration branch into the feature branch. The water calculator might be fixed by this.
3. Any suggestion would be appreciated 🤣 

Also, I addressed the [bug](https://lite-farm.atlassian.net/wiki/spaces/~63c1e5e3c52a63dcda83b94f/pages/1287094313/Unit+displays+instances?focusedCommentId=1292140549) you found. It would be great if you could review the last commit I made.
Thank you! :)

Duncan-Brain commented:
@SayakaOno 

Yeah... naming is so hard! It was just a thing I noticed but maybe another day haha

The bug: Rounding is complex so I wasn't sure it was an actual bug or just how we want rounding to be. That original issue I saw no longer exists there. Even now with the fix changing a different unit... the actual seed weight units over-rounds it too. Using:  5 for # of plants, and anything for plant spacing, estimated seed hidden value = 0.0425 kg > 40mt > 0.04 kg

Duncan-Brain reviewed (approved):
Makes sense -- ready for testing!
